### PR TITLE
Correctly transform the "Origin" header

### PR DESCRIPTION
**Pull request by LavrovArtem**
/cc @inikulin @churkin
@VasilyStrelyaev check test name

### CODE REVIEW

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
lgtm

File: test/server/proxy-test.js [hunk 1]
```diff
@@ -122,7 +122,7 @@ describe('Proxy', function () {
             res.end();
         });
 
-        app.get('/B234325/reply-with-origin', function (req, res) {
+        app.get('/xhr/reply-with-origin', function (req, res) {
```
Comment (inikulin) [inline test/server/proxy-test.js]: Let's stick with previously used notation for test routes:
```js
app.get('/B234325,GH-284/reply-with-origin', function (req, res) {
...
});
```

**Comment by inikulin** [comment]
\r-

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

File: test/server/proxy-test.js [hunk 2]
```diff
@@ -665,5 +665,22 @@ describe('Proxy', function () {
                 done();
             });
         });
+
+        it('Should correct transform the "Origin" header for request without the "Referer" header (GH-284)', function (done) {
```
Comment (VasilyStrelyaev) [inline test/server/proxy-test.js]:
```
        it('Should transform the "Origin" header for requests without the "Referer" header correctly (GH-284)', function (done) {
```

**Comment by LavrovArtem** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by inikulin** [comment]
lgtm

**Comment by churkin** [comment]
What about iframes?

**Comment by inikulin** [comment]
@churkin What kind of problem with iframes we may have here?

**Comment by churkin** [comment]
For example, in the iframe without src `window.location` can be 'about:blank'

**Comment by LavrovArtem** [comment]
FPR

File: src/client/utils/dom.js [hunk 3]
```diff
@@ -129,6 +129,14 @@ export function getParentWindowWithSrc (window) {
     return getParentWindowWithSrc(parent);
 }
 
+export function getWindowWithSrc (window) {
```
Comment (inikulin) [inline src/client/utils/dom.js]: What if we are in iframe without `src` which is in iframe without `src`?

File: src/client/sandbox/xhr/index.js [hunk 4]
```diff
@@ -89,6 +90,7 @@ export default class XhrSandbox extends SandboxBase {
             // server side. So, we pass the CORS support flag to inform the proxy that it can analyze the
             // Access-Control_Allow_Origin flag and skip "preflight" requests.
             xhr.setRequestHeader(XHR_HEADERS.requestMarker, 'true');
+            xhr.setRequestHeader(XHR_HEADERS.referer, getWindowWithSrc(window).location.toString());
```
Comment (inikulin) [inline src/client/sandbox/xhr/index.js]: We need client tests for this

**Comment by inikulin** [comment]
\r-

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

File: src/request-pipeline/context.js [hunk 5]
```diff
@@ -128,6 +128,9 @@ export default class RequestPipelineContext {
                 this.dest.reqOrigin = urlUtils.getDomain(parsedReferer.dest);
             }
 
+            if (!this.dest.reqOrigin && this.req.headers[XHR_HEADERS.origin])
```
Comment (inikulin) [inline src/request-pipeline/context.js]: Why not 
```js
else if( this.req.headers[XHR_HEADERS.origin])
```
?

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

File: src/request-pipeline/context.js [hunk 6]
```diff
@@ -126,7 +126,8 @@ export default class RequestPipelineContext {
             if (parsedReferer) {
                 this.dest.referer   = parsedReferer.dest.url;
                 this.dest.reqOrigin = urlUtils.getDomain(parsedReferer.dest);
-            }
+            } else if (this.req.headers[XHR_HEADERS.origin])
```
Comment (inikulin) [inline src/request-pipeline/context.js]: Please, place `else if` on the newline. Odd, I thought we've had eslint rule for this

**Comment by inikulin** [comment]
\r-

**Comment by LavrovArtem** [comment]
FPR

**Comment by inikulin** [comment]
lgtm

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

**Comment by inikulin** [comment]
We need client tests.
\r-

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

File: test/client/fixtures/sandbox/xhr-test.js [hunk 7]
```diff
@@ -82,3 +93,39 @@ asyncTest('parameters must pass correctly in xhr event handlers (T239198)', func
     request.open('GET', '/xhr-large-response', true);
     request.send(null);
 });
+
+if (!browserUtils.isIE9) {
+    asyncTest('send correct the origin header (GH-284)', function () {
```
Comment (VasilyStrelyaev) [inline test/client/fixtures/sandbox/xhr-test.js]:
```
    asyncTest('send the origin header correctly (GH-284)', function () {
```

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by inikulin** [comment]
lgtm

**Comment by churkin** [comment]
lgtm

**Pull request merged**